### PR TITLE
DEV: Allow Ember CLI assets to be used by development Rails app

### PR DESCRIPTION
**Pull request by davidtaylorhq**
Previously, accessing the Rails app directly in development mode would give you assets from our 'legacy' Ember asset pipeline. The only way to run with Ember CLI assets was to run ember-cli as a proxy. This was quite limiting when working on things which are bypassed when using the ember-cli proxy (e.g. changes to `application.html.erb`). Also, since `ember-auto-import` introduced chunking, visiting `/theme-qunit` under Ember CLI was failing to include all necessary chunks.

This commit teaches Sprockets about our Ember CLI assets so that they can be used in development mode, and are automatically collected up under `/public/assets` during `assets:precompile`. As a bonus, this allows us to remove all the custom manifest modification from `assets:precompile`.

The key changes are:
- Introduce a shared `EmberCli.enabled?` helper
- When ember-cli is enabled, add ember-cli `/dist/assets` as the top-priority Rails asset directory
- Have ember-cli output a `chunks.json` manifest, and teach `preload_script` to read it and append the correct chunks to their associated `afterFile`
- Remove most custom ember-cli logic from the `assets:precompile` step. Instead, rely on Rails to take care of pulling the 'precompiled' assets into the `public/assets` directory. Move the 'renaming' logic to runtime, so it can be used in development mode as well.
- Remove fingerprinting from `ember-cli-build`, and allow Rails to take care of things

Long-term, we may want to replace Sprockets with the lighter-weight Propshaft. The changes made in this commit have been made with that long-term goal in mind.

tldr: when you visit the rails app directly, you'll now be served the current ember-cli assets. To keep these up-to-date make sure either `ember serve`, or `ember build --watch` is running. If you really want to load the old non-ember-cli assets, then you should start the server with `EMBER_CLI_PROD_ASSETS=0`. (the legacy asset pipeline will be removed very soon)

### CODE REVIEW

File: lib/ember_cli.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    case name
+    when "application"
+      "discourse"
+    when "discourse/tests/test-support-rails"
+      "test-support"
+    when "discourse/tests/test-helpers-rails"
+      "test-helpers"
+    else
+      name
+    end
```
Comment (tgxworld) [inline lib/ember_cli.rb:41]:
```suggestion
    {
      "application": "discourse",
      "discourse/tests/test-support-rails": "test-support",
      "discourse/tests/test-helpers-rails": "test-helpers",
    }[name] || name
```

I wonder if using a hash is better here and avoid unnecessary comparisons.

Comment (tgxworld) [reply lib/ember_cli.rb:41]: Might also want to consider moving the hash to a constant and avoid an allocation each time the method is called.

File: lib/ember_cli.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+
+module EmberCli
+  def self.enabled?
+    ENV["EMBER_CLI_PROD_ASSETS"] != "0"
+  end
+
+  def self.script_chunks
+    return @@chunk_infos if defined? @@chunk_infos
```
Comment (tgxworld) [inline lib/ember_cli.rb:9]: Based on what I'm seeing, this info here is computed for every unicorn process the first time this method is called. Would it be helpful to preload this information before forking?

File: config/application.rb [hunk 3]
```diff
@@ -272,6 +282,21 @@ def watchable_args
     Sprockets.register_mime_type 'application/javascript', extensions: ['.js', '.es6', '.js.es6'], charset: :unicode
     Sprockets.register_postprocessor 'application/javascript', DiscourseJsProcessor
 
+    # The sprockets sourcemap pipeline overrites our ember-cli-generated sourcemaps, so let's disable it
+    # (it doesn't work with our DiscourseJsProcessor anyway)
+    config.assets.configure do |env|
+      env.cache = nil # TODO REMOVE THIS
```
Comment (tgxworld) [inline config/application.rb:288]: I think it'll be helpful if we can elaborate further on the comment here and why there is a need to remove this.

File: config/application.rb [hunk 4]
```diff
@@ -272,6 +282,21 @@ def watchable_args
...
+      env.unregister_processor('application/javascript', Sprockets::Preprocessors::DefaultSourceMap)
+      env.unregister_postprocessor('application/javascript', Sprockets::Rails::SourcemappingUrlProcessor)
+    end
+
+    if EmberCli.enabled?
+      Discourse::Application.initializer :prepend_ember_assets do |app|
+        # Needs to be in its own initializer so it runs after Sprockets' append_assets_path
```
Comment (tgxworld) [inline config/application.rb:295]: Is `append_assets_path` an initializer defined in sprockets?

**Review by tgxworld — APPROVED** [review]
Some comments but nothing blocking and the changes look good to me.

File: config/application.rb (hunk 4 above)
Comment (davidtaylorhq) [reply config/application.rb:295]: Yes. I'll make the comment clearer :+1:

File: lib/ember_cli.rb [hunk 5]
```diff
@@ -1,6 +1,12 @@
 # frozen_string_literal: true
 
 module EmberCli
+  ALIASES ||= {
```
Comment (tgxworld) [inline lib/ember_cli.rb:4]:
```suggestion
  ALIASES = {
```

Our code base is littered with `||=` when assigning constants but we actually have no good reason to do that.

Comment (davidtaylorhq) [reply lib/ember_cli.rb:4]: I think I normally get comments which say the opposite, and ask me to change `=` to `||=` 😅 

But I agree, I can't think of any good reason, so let's try without and see if it triggers any warnings.

**Comment by davidtaylorhq** [comment]
(Updated to work with Sprockets 3, since we decided not to go ahead with the upgrade to 4)
